sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentUtils.scala

```diff
@@ -0,0 +1,275 @@
+          conf, addError, colPath)
+        resolvedValue.getOrElse(col)
+
+      case (valueType: MapType, colType: MapType) =>
+        TableOutputResolver.checkNullability(value, col, conf, addError, colPath)
+        val resolvedValue = TableOutputResolver.resolveMapType(
+          value, valueType, colType,
+          conf, addError, colPath)
+        resolvedValue.getOrElse(col)
+
+      case (valueType, colType) =>
+        TableOutputResolver.checkNullability(value, col, conf, addError, colPath)
+
+        val colTypeHasCharVarchar = CharVarcharUtils.hasCharVarchar(colType)
+        val colTypeWithoutCharVarchar = if (colTypeHasCharVarchar) {
+          CharVarcharUtils.replaceCharVarcharWithString(colType)
+        } else {
+          colType
+        }
+
+        validateAssignment(valueType, colTypeWithoutCharVarchar, addError, colPath)
+
+        val casted = TableOutputResolver.cast(
+          value, colTypeWithoutCharVarchar,
+          conf, colPath.quoted)
+
+        if (conf.charVarcharAsString || !colTypeHasCharVarchar) {
+          casted
+        } else {
+          CharVarcharUtils.stringLengthCheck(casted, colType)
+        }
+    }
+  }
+
+  private def validateAssignment(
+      valueType: DataType,
+      expectedType: DataType,
+      addError: String => Unit,
+      colPath: Seq[String]): Unit = {
+
+    conf.storeAssignmentPolicy match {
+      case StoreAssignmentPolicy.STRICT | StoreAssignmentPolicy.ANSI =>
+        DataType.canWrite(
+          valueType, expectedType, byName = true, conf.resolver, colPath.quoted,
+          conf.storeAssignmentPolicy, addError)
+
+      case _ => // OK
+    }
+  }
+
+  /**
+   * Checks whether assignments are aligned and are compatible with table columns.
+   *
+   * @param attrs table attributes
+   * @param assignments assignments to check
+   * @return true if the assignments are aligned
+   */
+  def aligned(attrs: Seq[Attribute], assignments: Seq[Assignment]): Boolean = {
+    if (attrs.size != assignments.size) {
+      return false
+    }
+
+    attrs.zip(assignments).forall { case (attr, assignment) =>
+      val key = assignment.key
+      val value = assignment.value
+
+      val attrType = CharVarcharUtils.getRawType(attr.metadata).getOrElse(attr.dataType)
+
+      sameRef(toRef(key), toRef(attr)) &&
+        DataType.equalsIgnoreCompatibleNullability(value.dataType, attrType) &&
+        (attr.nullable || !value.nullable)
+    }
+  }
+
+  private def sameRef(ref: Seq[String], otherRef: Seq[String]): Boolean = {
+    ref.size == otherRef.size && ref.zip(otherRef).forall { case (namePart, otherNamePart) =>
+      conf.resolver(namePart, otherNamePart)
+    }
+  }
+
+  private def toRef(expr: Expression): Seq[String] = expr match {
+    case attr: AttributeReference =>
+      Seq(attr.name)
+    case Alias(child, _) =>
+      toRef(child)
+    case GetStructField(child, _, Some(name)) =>
+      toRef(child) :+ name
+    case other: ExtractValue =>
```

Comment:
how about `GetArrayStructField`?

Reply:
For ALTER COLUMN we support a special syntax to reference any inner field, for example, `array_col.element.field1`, `map_col.key.field2`, etc. Shall we support this syntax in UPDATE as well? The related code is in `StructType.findNestedField`

Reply:
We should eventually. This PR doesn't support updating arrays or maps, though. I wanted to work on it later and unblock further row-level operation development for now.

Reply:
Actually, I can add support for those expressions here but fail temporary in the rewrite logic.